### ISSUE TITLE
Garbage appended to end of files

### ISSUE THREAD

**Issue by 89z** [joshbarrass/deezerdl#6]
If I use this project:

https://github.com/moon004/Go-deezer-downloader

and download MP3_320, I can remux with FFmpeg:

~~~
ffmpeg -i in.mp3 out.mp3
~~~

However if I get the same file with this project:

~~~
deezerdl download track 75498415 -f MP3_320
~~~

If I try FFmpeg, I get this result:

~~~
[mp3float @ 0000025f4ed6a140] Header missing
Error while decoding stream #0:0: Invalid data found when processing input
[libmp3lame @ 0000025f4ed40400] Trying to remove 1152 samples, but the queue is
empty
~~~

**Comment by 89z** [comment]
Looks like the error is here:

https://github.com/joshbarrass/deezerdl/blob/9df60fd906641553561a7ab078834c9fca578669/pkg/deezer/blowfish.go#L61

you are assuming that the input file will be a `fileChunkSize`, when usually the last read slice will be smaller. You need to detect that somehow and resize the buffer like this:

~~~go
buf = buf[:something]
~~~

its difficult to detect the end if you are reading from a `Reader`, might be easier to read the whole file into a slice and iterate like that. I have an implementation here:

https://github.com/89z/deezer/blob/v1.0.6/deezer.go#L34-L51

So in short, your created files are correct, but they have extra bytes on the end which causes errors for FFmpeg and maybe other programs.

**Comment by joshbarrass** [comment]
This is a little redundant at this point, as this project no longer works (at least not for downloading FLAC files) due to changes in deezer. I will look into fixing this at some point, but since this program is a little useless now this may not be soon.

In the mean time, whilst ffmpeg struggles due to the added garbage, the `flac` command line utility does not and can (for FLAC files at least) remove the garbage by re-encoding the file. So ffmpeg can be used to test for the presence of any garbage and a command like `flac -8 -V -f file.flac` can be used to repair them.

**Comment by 89z** [comment]
The issue was never about FLAC...

**Comment by joshbarrass** [comment]
I'm well aware, but the issue affects most files downloaded by this tool. For example, it presents itself in FLAC files as invalid frames and causes similar errors in ffmpeg. Fixing already-downloaded MP3 files would have to be done manually to avoid lossy->lossy re-encoding, but FLAC files, being lossless, can be repaired simply by re-encoding them, which has no effect on the sound in the file. My reply was simply offering a solution for anyone else whose FLAC files may have been affected by the same bug until I get time to draft a solution.

You're welcome to submit a pull request yourself to fix this bug.

**Comment by joshbarrass** [comment]
Fixed as of 34f373d5f752f36c483bf4d84566daf4ab6938df

`File.Read` reports the number of bytes read, so the bytes written can be truncated easily with this. It's also enforced now that bytes read must equal bytes written, else the program throws an error.
